### ISSUE TITLE
Foreach Functions Tracking Issue

### ISSUE THREAD

**Issue by crcrpar** [pytorch/pytorch#58833]
Foreach functions would improve optimizers performance as described in https://github.com/pytorch/pytorch/issues/38655, but they needs fixes & improvements to replace the current optimizers implementation.

# A: Functionality enhancement
To make is possible to use foreach functions in optimizers, we need to support tensors on different devices, dtypes, and layouts even if those inputs are not going through fastpath (Multi Tensor Apply).

- [ ] Push inputs incompatible with MTA to slowpath - #56993 
- [ ] Refactor unittest
    - [ ] unary
    - [ ] binary
    - [ ] pointwise
    - [ ] min/max-imum
- [ ] Support implicit broadcasting via slowpath

# B: Performance improvement

- [ ] Type promotion in fastpath.
    - [ ] Unary
    - [ ] Binary
    - [ ] Pointwise
- [ ] Optimizers
    - [ ] Rewrite `torch.optim._multi_tensor` optimizers in functional form.
- [ ] Separate input `TensorList`s by devices & dtypes into bunch of vectors of Tensors in C++ to utilize MTA kernel as much as possible.
- [ ] Let binary fastpath support inputs of tensorlist and complex scalarlist.
    - binary foreach's with two tensorlists and one tensorlist and one scalar support complex. The limitation derives from that [`TensorListScalarListMetadata`](https://github.com/pytorch/pytorch/blob/50ded095e41d6d3edcde9e5e4a608b89b9a374ce/aten/src/ATen/native/cuda/MultiTensorApply.cuh#L40) violates kernel argument size limit when `scalar_vals_t` is `c10::complex<double>` and `n` is `1`.
    - https://github.com/pytorch/pytorch/blob/50ded095e41d6d3edcde9e5e4a608b89b9a374ce/aten/src/ATen/native/cuda/ForeachBinaryOpList.cu#L20
    - https://github.com/pytorch/pytorch/blob/50ded095e41d6d3edcde9e5e4a608b89b9a374ce/aten/src/ATen/native/cuda/ForeachBinaryOpScalar.cu#L19
    - https://github.com/pytorch/pytorch/blob/50ded095e41d6d3edcde9e5e4a608b89b9a374ce/aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu#L20

For Binary & Pointwise support of type promotion, I think we need a new macro that can handle multiple C++ types or one that can be nested like [APEX](https://github.com/NVIDIA/apex/blob/082f999a6e18a3d02306e27482cc7486dab71a50/csrc/type_shim.h).

# C: autograd
See https://github.com/pytorch/pytorch/issues/53796.

---
Please kindly let me know if I miss anything and/or I make some mistakes
cc: @ptrblck @mcarilli @ngimel

**Comment by zou3519** [comment]
Marking as triage review because I don't know what labels to add to this. @ngimel I'm not sure what to label this as. We could label it as `module: optim` or `module: performance`, or alternative add some new labels (`module: MTA` or `module: foreach`?)

**Comment by ngimel** [comment]
@zou3519 We could add mta label and also performance.

**Comment by ngimel** [comment]
Thank you, @crcrpar, this is a good plan. You might also add reduction mta kernels to "Performance Improvement" section - they'd help with some optimizers and gradient clipping.

**Comment by UmaisZahid** [comment]
Are there any plans of adding _foreach_max? (Not element-wise maximum)

**Comment by jisaacso** [comment]
> Are there any plans of adding _foreach_max? (Not element-wise maximum)

cc @mikaylagawarecki

**Comment by crcrpar** [comment]
https://github.com/pytorch/pytorch/pull/93995

**Comment by crcrpar** [comment]
- [ ] try to generate `THP<codegen'd Foreach Backward TraceableFunction_properties[]` so that `result_tensors[0].grad_fn` will be explicit
- [ ] avoid having each input Tensor's grad_fn have the reference to redundant tensors which seems added by the backward backward function struct

**Comment by vadimkantorov** [comment]
a request for _foreach_copy_: https://github.com/pytorch/pytorch/issues/90539

**Comment by vadimkantorov** [comment]
Related:
- https://github.com/pytorch/pytorch/issues/100249

**Comment by crcrpar** [comment]
that sounds like a nice point...

**Comment by crcrpar** [comment]
I think major remaining tasks to make foreach feature complete-ish would be
- [ ] forward mode AD of out-place foreach functions. this could be trickier than in-place.
- [ ] Documentation. I've already done unary functions but not for the other foreach functions.

**Comment by vadimkantorov** [comment]
> Related (especially if backward for foreach functions starts getting implemented):
> 
> * [[discussion] "TensorList" as first-class abstraction (including python frontend) and as key for dispatch for merging `torch._foreach_*` into regular `torch.*` functions #100249](https://github.com/pytorch/pytorch/issues/100249)

Was the idea of merging foreach and non-foreach functions by adding overloads and dispatching by TensorList considered? Any outcomes of this discussion? Once the foreach funcions are added in the docs, it would take extra further effort to remove them... And if the overloads of foreach functions are roughly matching the signatures of regular `torch.*` functions, can't they be just added as overloads to `torch.*`?
